METRON-1304: Allow metron-bro-plugin-kafka to include or exclude logs

JonZeolla opened the pull request:
This allows us the option to specify the inclusion (either all active logs or only specified logs) and exclusion of logs to send to kafka.  If `Kafka::logs_to_send` is not set and `send_all_active_logs` is not overridden, no logs are sent to kafka, regardless of the value of `Kafka::logs_to_exclude`.

## Testing
An easy way to test this is to use `bro-pkg` to install the plugin (which runs all of the tests by default), or you can go on a properly configured box (wrt `btest`) and run `cd metron-bro-plugin-kafka/tests/ && btest -d`.

## Outstanding items:
- Add btests for:
  - [x] No logs should be sent when `logs_to_send` and `logs_to_exclude` are empty.
  - [x] Only the specified logs should be sent when `logs_to_send` is set but `logs_to_exclude` is empty
  - [x] No logs should be sent when `logs_to_send` is empty and `logs_to_exclude` is set.
  - [x] Only the logs both in `logs_to_send` and not in `logs_to_exclude` should be sent when both `logs_to_send`  and `logs_to_exclude` are set.
  - [x] All logs but the blacklisted logs should be sent when `send_all_active_logs` is true but `logs_to_exclude` contains some active logs.
- Documentation
  - [X] Add `logs_to_exclude` to the README
- [x] Test against the apache/metron full-dev.

ottobackwards commented:
logs_to_exclude should be in the readme, and not just the bro file

JonZeolla commented:
Yup, I'm typing that up as we speak.

nickwallen commented:
@JonZeolla Can you send an update to the Dev list on where we are at in migrating the Bro plugin?  I had not even realized that you had added the code to this repo.  

We also still have the Bro plugin code in the main Metron repository.  Don't we have some clean-up to do before we start merging new changes?

JonZeolla commented:
Yeah, I did send an initial email out [last night](https://lists.apache.org/thread.html/621cbeca432fef0170836e07036e309d943068c5d6a544c1ef2136f9@%3Cdev.metron.apache.org%3E), and in a previous email in that thread I outlined a suggested roadmap, but I will bring it up and clarify things so we can discuss on the list instead of here.  Thanks

nickwallen commented:
My bad, you did.  You're totally on top of it.  Ignore me.  :)

nickwallen commented on scripts/Bro/Kafka/logs-to-kafka.bro:
```diff
@@ -14,32 +14,37 @@
-				$config = table(["stream_id"] = fmt("%s", stream_id))
-			];
+		if ( stream_id in Kafka::logs_to_exclude ||
+		    (|Kafka::logs_to_send| > 0 && stream_id !in Kafka::logs_to_send) )
```
Why do we have to check that `logs_to_send` > 0 ? Is this necessary before doing a 'contains' (`in`)?

If it is necessary then we should do the same for `logs_to_exclude`.  If it is NOT necessary, then let's just get rid of it to simplify the logic.

nickwallen replied:
If a user configures nothing, neither defines `logs_to_send` nor `logs_to_exclude`, then we expect ALL logs to go to Kafka?

JonZeolla replied:
Yeah, that's valid, I have removed the check and simplify.

Yeah, I would prefer a default 'send everything' policy when someone loads the package, as long as it's otherwise configured.  That said, it will require a bit of Metron testing to make sure that it can handle that.  We don't currently handle some of the less interesting logs that are on by default, like packet filter or loaded scripts.

JonZeolla replied:
Actually, wait, sorry.  If `|Kafka::logs_to_send| > 0` is removed, this doesn't send when `logs_to_send` is unset.  Re-adding this.

JonZeolla replied:
I was talking to a few people in the bro community about this and I'm hearing that people mostly prefer an unset send_logs to send nothing.  What are your thoughts on that?  I would prefer to send all by default, but it's not a huge deal to go either way for me.

nickwallen replied:
I too would rather an unset `logs_to_send` to send nothing.  I'd rather not 'surprise' the user with more logs.  Especially with how difficult it can be to truly 'see' the variety of logs that are actually landing in a topic.

JonZeolla replied:
Ok I'm convinced, I guess my posture on this is more aggressive than most.  I will adjust

nickwallen commented on scripts/init.bro:
```diff
@@ -18,11 +18,20 @@
+	## Maximum wait on shutdown in milliseconds
+	const max_wait_on_shutdown: count = 3000 &redef;
+
+	## Boolean to JSON with a log stream identifier
```
Thanks for adding all the comments.    But this one doesn't read so well.  Or maybe I just don't read so well. :)  Read this one over once more.  If it still makes sense to you, then I'm good with it.

JonZeolla replied:
Lol.  I will update, thanks

JonZeolla commented:
This is ready for review.  I refactored and added tests for the new include/exclude functionality and also found out that the unit tests are broken for `0.2`, which this fixes.

JonZeolla commented:
An easy way to test this is to use bro-pkg to install the plugin from my branch, or you can go on a properly configured bro box to the `tests` directory and run `btest -d`.

ottobackwards commented on README.md:
```diff
@@ -11,6 +11,32 @@ This software is a part of the [Apache Metron](http://metron.apache.org/) projec
+### `bro-pkg` Installation
+
+`bro-pkg` is the preferred mechanism for installing this plugin, as it will dynamically retrieve, build, test, and load the plugin.  Note, that you will still need to [activate](#activation) and configure the plugin after your installation.
+
```
Do these libs have to be installed in any certain place?  Relative to the Bro Install?

JonZeolla replied:
There is a suggestion via a [user var](https://bro-package-manager.readthedocs.io/en/stable/package.html#user-vars-field) defined [here](https://github.com/apache/metron-bro-plugin-kafka/blob/d1a09b6a50f20e5fa6cf5c758eea8b0d39ce65be/bro-pkg.meta#L16) but it can be overridden during install.  Whatever you provide gets passed into the [build_command](https://github.com/apache/metron-bro-plugin-kafka/blob/d1a09b6a50f20e5fa6cf5c758eea8b0d39ce65be/bro-pkg.meta#L5).

ottobackwards replied:
I mean do we have to document that the libs have to be installed to where bro looks for libs?  Like /usr/local vs /opt

JonZeolla replied:
The libs don't need to be installed anywhere in particular, I could put them in my home directory.

ottobackwards replied:
ok

JonZeolla replied:
As a part of the review process, this improvement was split into #17

ottobackwards commented on README.md:
```diff
@@ -11,6 +11,32 @@ This software is a part of the [Apache Metron](http://metron.apache.org/) projec
 ## Installation
 
+### `bro-pkg` Installation
+
```
What is the case for having more than one way to install?  When would you not want to use bro-pkg?

JonZeolla replied:
Added some context to the README.

JonZeolla replied:
As a part of the review process, this improvement was split into #17

ottobackwards commented on README.md:
````diff
@@ -42,22 +68,47 @@ This software is a part of the [Apache Metron](http://metron.apache.org/) projec
+ * Defining `logs_to_send` will send the HTTP and DNS records to the brokers specified in your `Kafka::kafka_conf`.
 ```
 @load packages/metron-bro-plugin-kafka/Apache/Kafka
 redef Kafka::logs_to_send = set(HTTP::LOG, DNS::LOG);
````
The example should be for multiple brokers, not just one since that will be the most common case won't it?

JonZeolla replied:
Sure, I added multiple brokers to example 1.

JonZeolla replied:
As a part of the review process, this improvement was split into #17

ottobackwards commented on README.md:
```diff
@@ -144,23 +194,35 @@ event bro_init() &priority=-5
 
 #### Notes
  * `logs_to_send` is mutually exclusive with `$pred`, thus for each log you want to set `$pred` on, you must individually setup a `Log::add_filter` and refrain from including that log in `logs_to_send`.
+ * In Bro 2.5.x the bro project introduced a [logger function](https://www.bro.org/sphinx/cluster/index.html#logger) which removes the logging functions from the manager thread, and taking advantage of that is highly recommended.  If you are running this plugin on Bro 2.4.x, you may encounter issues where the manager thread is taking on too much responsibility and pinning a single CPU core withou ...
```
s/lesson/lessen/

JonZeolla replied:
As a part of the review process, this improvement was split into #17

JonZeolla commented:
@ottobackwards That last commit should have addressed all of your feedback.

JonZeolla commented:
If you're looking to test this by spinning up full-dev in apache/metron without skipping the sensors tag (a non-default setting), you would need to merge in apache/metron#1238.  There are other ways to test this without doing that, but I assume that's a typical way people may want to look at this.

JonZeolla commented:
Based on the recent release discussions I'd like to get this in.  @nickwallen do you have any additional feedback?

nickwallen commented:
@JonZeolla Can you add the latest testing instructions in the description?  This seems to have evolved over time and am not sure what is needed at this point.

nickwallen commented:
@JonZeolla I just want to confirm, all of this is related only to include/exclude functionality?  There seems to be more here, but maybe I am misunderstanding on first glance.

JonZeolla commented:
https://github.com/apache/metron-bro-plugin-kafka/pull/2#issuecomment-429502420 is still accurate for testing.  I'll edit the initial comment to add it.

JonZeolla commented:
In looking through the PR, this is mostly related to the include/exclude functionality yes, but you are right that it may look like there are some other things included because:
* As a part of updating the documentation for these new features I attempted to make things more uniform in format (which included some text reordering), and I added details regarding what a successful `bro-pkg` install looks like (before this PR that was broken/undocumented)
* This also fixes the previously broken btest which was a 1 line change to `tests/Baseline/kafka.show-plugin/output`
* There are some minor documentation changes such as changing suggested priority to -10, or adding multiple brokers to the docs in response to a [review comment](https://github.com/apache/metron-bro-plugin-kafka/pull/2/files#r224996990).

nickwallen commented:
@JonZeolla Yes, I would definitely appreciate if you could break this into separate, focused PRs.

This PR itself has been open since Nov 2017 and it seems like it has been accumulating changes.  Would be better to just open a PR when you're ready for a review. 

You might consider closing some of the other ones that you have open (#12, #8, ?), since those don't seem to be ready for review.  It makes it difficult to track what needs reviewed if we have PRs just hanging out there.

JonZeolla commented:
Good to go on this one @nickwallen

JonZeolla commented:
Note that this one now depends on apache/metron-bro-plugin-kafka#16 to work properly because I split out the btest bugfix.

JonZeolla commented:
Suggested order of review:  #16 -> #2 -> #17 -> #13

nickwallen commented:
I am getting a test failure.
```
[root@localhost ~]# bro-pkg install http://github.com/jonzeolla/metron-bro-plugin-kafka --version=METRON-1304
The following packages will be INSTALLED:
  http://github.com/jonzeolla/metron-bro-plugin-kafka (METRON-1304)

Verify the following REQUIRED external dependencies:
(Ensure their installation on all relevant systems before proceeding):
  from http://github.com/jonzeolla/metron-bro-plugin-kafka (METRON-1304):
    librdkafka ~0.9.4

Proceed? [Y/n] Y
http://github.com/jonzeolla/metron-bro-plugin-kafka asks for LIBRDKAFKA_ROOT (Path to librdkafka installation tree) ? [/usr/local/lib]
Saved answers to config file: /root/.bro-pkg/config
Running unit tests for "http://github.com/jonzeolla/metron-bro-plugin-kafka"
[ 90%] kafka.show-plugin ... failed
  % 'btest-diff output' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  Apache::Kafka - Writes logs to Kafka (dynamic, version 0.3)
      [Writer] KafkaWriter (Log::WRITER_KAFKAWRITER)
      [Constant] Kafka::kafka_conf
      [Constant] Kafka::topic_name
      [Constant] Kafka::max_wait_on_shutdown
      [Constant] Kafka::tag_json
      [Constant] Kafka::json_timestamps
      [Constant] Kafka::debug

  == Diff ===============================
  --- /tmp/test-diff.22832.output.baseline.tmp	2018-11-09 22:51:27.714922560 +0000
  +++ /tmp/test-diff.22832.output.tmp	2018-11-09 22:51:27.719922348 +0000
  @@ -4,5 +4,6 @@
       [Constant] Kafka::topic_name
       [Constant] Kafka::max_wait_on_shutdown
       [Constant] Kafka::tag_json
  +    [Constant] Kafka::json_timestamps
       [Constant] Kafka::debug

  =======================================

  % cat .stderr

1 of 10 tests failed
error: http://github.com/jonzeolla/metron-bro-plugin-kafka tests failed, inspect contents of /root/.bro-pkg/testing/metron-bro-plugin-kafka for details
Proceed to install anyway? [N/y]
```

JonZeolla commented:
Yes that's why I said it depends on #16 and mentioned the order of review should be #16 -> #2 -> #17 -> #13

JonZeolla commented:
Okay, pushed that fix and merged in master.

nickwallen commented:
+1 Looks good @JonZeolla Thanks
